lib/models/asyncapi.js

```diff
@@ -234,6 +234,103 @@ class AsyncAPIDocument extends Base {
   traverseSchemas(callback, schemaTypesToIterate) {
     traverseAsyncApiDocument(this, callback, schemaTypesToIterate);
   }
+
+  /**
+   * Converts a valid AsyncAPI document to a JavaScript Object Notation (JSON) string.
+   * A stringified AsyncAPI document using this function should be parsed via the AsyncAPIDocument.parse() function - the JSON.parse() function is not compatible.
+   * 
+   * @param {AsyncAPIDocument} doc A valid AsyncAPIDocument instance.
+   * @param {(number | string)=} space Adds indentation, white space, and line break characters to the return-value JSON text to make it easier to read.
+   * @returns {string}
+   */
+  static stringify(doc, space) {
+    return JSON.stringify(doc.json(), refReplacer(), space);
+  }
+
+  /* eslint-disable sonarjs/cognitive-complexity */
+  /**
+   * Converts a valid stringified AsyncAPIDocument instance into an object.
+   * 
+   * @param {string} doc A valid stringified AsyncAPIDocument instance.
+   * @returns {AsyncAPIDocument}
+   */
+  static parse(doc) { // NOSONAR
+    const objToPath = new Map();
+    const pathToObj = new Map();
+    const parsedJSON = JSON.parse(doc);
+    
+    function traverse(parent, field) {
+      let objOrPath = parent;
+      let path = '$ref:$';
+  
+      if (field !== undefined) {
+        // here can be string with `$ref` prefix or normal value 
+        objOrPath = parent[String(field)];
+        const concatenatedPath = field ? `.${  field}` : '';
+        path = objToPath.get(parent) + (Array.isArray(parent) ? `[${field}]` : concatenatedPath);
+      }
+  
+      objToPath.set(objOrPath, path);
+      pathToObj.set(path, objOrPath);
+      
+      const ref = pathToObj.get(objOrPath);
+      if (ref) {
+        parent[String(field)] = ref;
+      }
+      if (objOrPath === '$ref:$' || ref === '$ref:$') {
+        parent[String(field)] = parsedJSON;
+      }
+  
+      // traverse all keys, only if object is array/object
+      if (objOrPath === Object(objOrPath)) {
+        for (const f in objOrPath) {
+          traverse(objOrPath, f);
+        }
+      }
+    }
+  
+    traverse(parsedJSON);
+    return new AsyncAPIDocument(parsedJSON);
+  }
+  /* eslint-enabled sonarjs/cognitive-complexity */
+}
+
+/**
+ * Replacer function (that transforms the result) for AsyncAPI.stringify() function.
+ * Handles circular references by replacing it by JSONPath notation. 
+ * 
+ * @private
+ */
+function refReplacer() {
```

Comment:
Interesting 🤔 

So instead of just printing `Circular` you actually print the reference to the location of the circular schema? 🤯

Reply:
I cannot print `Circular` (or annotate object as circular) due to fact that I need the whole object after parsing :) There is a also deeper problem, because if in one place given object is circular, in another it don't be, so I cannot also change the references to this object from POV of it parent 😄  Yeah, it's very hard to understand, if you don't, please write, I will write example.